// @squircle-js/react/dist/index.js
"use strict";
"use client";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  Squircle: () => Squircle,
  SquircleNoScript: () => SquircleNoScript,
  StaticSquircle: () => StaticSquircle
});
module.exports = __toCommonJS(src_exports);
var import_figma_squircle2 = require("figma-squircle");
var import_react5 = require("react");

// src/use-element-size.ts
var import_react3 = require("react");

// src/use-event-listener.ts
var import_react2 = require("react");

// src/use-isomorphic-layout-effect.ts
var import_react = require("react");
var useIsomorphicLayoutEffect = typeof window !== "undefined" ? import_react.useLayoutEffect : import_react.useEffect;

// src/use-event-listener.ts
function useEventListener(eventName, handler, element, options) {
  const savedHandler = (0, import_react2.useRef)(handler);
  useIsomorphicLayoutEffect(() => {
    savedHandler.current = handler;
  }, [handler]);
  (0, import_react2.useEffect)(() => {
    var _a;
    const targetElement = (_a = element == null ? void 0 : element.current) != null ? _a : window;
    if (!(targetElement && targetElement.addEventListener))
      return;
    const listener = (event) => savedHandler.current(event);
    targetElement.addEventListener(eventName, listener, options);
    return () => {
      targetElement.removeEventListener(eventName, listener, options);
    };
  }, [eventName, element, options]);
}

// src/use-element-size.ts
function useElementSize(defaultSize) {
  var _a, _b;
  const [ref, setRef] = (0, import_react3.useState)(null);
  const [size, setSize] = (0, import_react3.useState)({
    width: (_a = defaultSize.defaultWidth) != null ? _a : 0,
    height: (_b = defaultSize.defaultHeight) != null ? _b : 0
  });
  const handleSize = (0, import_react3.useCallback)(() => {
    setSize({
      width: (ref == null ? void 0 : ref.offsetWidth) || 0,
      height: (ref == null ? void 0 : ref.offsetHeight) || 0
    });
  }, [ref == null ? void 0 : ref.offsetHeight, ref == null ? void 0 : ref.offsetWidth]);
  useEventListener("resize", handleSize);
  useIsomorphicLayoutEffect(() => {
    handleSize();
  }, [ref == null ? void 0 : ref.offsetHeight, ref == null ? void 0 : ref.offsetWidth]);
  return [setRef, size];
}

// src/no-js.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var SquircleNoScript = () => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("noscript", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("style", { type: "text/css", children: `[data-squircle] { clip-path: none !important; border-radius: attr(data-squircle) !important; }` }) });
};

// src/index.tsx
var import_react_slot2 = require("@radix-ui/react-slot");

// src/StaticSquircle.tsx
var import_react_slot = require("@radix-ui/react-slot");
var import_figma_squircle = require("figma-squircle");
var import_react4 = require("react");
var import_jsx_runtime2 = require("react/jsx-runtime");
var StaticSquircle = (_a) => {
  var _b = _a, {
    asChild,
    width,
    height,
    cornerRadius,
    cornerSmoothing,
    style
  } = _b, props = __objRest(_b, [
    "asChild",
    "width",
    "height",
    "cornerRadius",
    "cornerSmoothing",
    "style"
  ]);
  const Component = asChild ? import_react_slot.Slot : "div";
  const path = (0, import_react4.useMemo)(() => {
    return (0, import_figma_squircle.getSvgPath)({
      width,
      height,
      cornerRadius,
      cornerSmoothing
    });
  }, [width, height, cornerRadius, cornerSmoothing]);
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(Component, __spreadValues({ style: __spreadValues({ clipPath: `path('${path}')` }, style) }, props));
};

// src/index.tsx
var import_jsx_runtime3 = require("react/jsx-runtime");
function Squircle(_a) {
  var _b = _a, {
    cornerRadius,
    cornerSmoothing = 0.6,
    asChild,
    style,
    width: w,
    height: h,
    defaultWidth,
    defaultHeight
  } = _b, props = __objRest(_b, [
    "cornerRadius",
    "cornerSmoothing",
    "asChild",
    "style",
    "width",
    "height",
    "defaultWidth",
    "defaultHeight"
  ]);
  const Component = asChild ? import_react_slot2.Slot : "div";
  const [ref, { width, height }] = useElementSize({
    defaultWidth,
    defaultHeight
  });
  const actualWidth = w != null ? w : width;
  const actualHeight = h != null ? h : height;
  const path = (0, import_react5.useMemo)(() => {
    return (0, import_figma_squircle2.getSvgPath)({
      width: actualWidth,
      height: actualHeight,
      cornerRadius,
      cornerSmoothing
    });
  }, [actualWidth, actualHeight, cornerRadius, cornerSmoothing]);
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
    Component,
    __spreadProps(__spreadValues({}, props), {
      ref,
      style: __spreadProps(__spreadValues({}, style), {
        borderRadius: cornerRadius,
        width: w != null ? w : defaultWidth,
        height: h != null ? h : defaultHeight,
        clipPath: `path('${path}')`
      }),
      "data-squircle": cornerRadius
    })
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Squircle,
  SquircleNoScript,
  StaticSquircle
});
